db/links/operations/create.py

```diff
@@ -0,0 +1,67 @@
+from alembic.operations import Operations
+from alembic.migration import MigrationContext
+from sqlalchemy import MetaData
+
+from db.columns.base import MathesarColumn
+from db.constraints.utils import naming_convention
+from db.tables.operations.select import reflect_table_from_oid, reflect_tables_from_oids
+from db.tables.utils import get_primary_key_column
+
+
+def create_foreign_key_link(engine, schema, column_name, referrer_table_oid, referent_table_oid, unique_link=False):
+    with engine.begin() as conn:
+        referent_table = reflect_table_from_oid(referent_table_oid, engine, conn)
+        referrer_table = reflect_table_from_oid(referrer_table_oid, engine, conn)
+        primary_key_column = get_primary_key_column(referent_table)
+        metadata = MetaData(bind=engine, schema=schema, naming_convention=naming_convention)
+        opts = {
+            'target_metadata': metadata
+        }
+        ctx = MigrationContext.configure(conn, opts=opts)
+        op = Operations(ctx)
+        column = MathesarColumn(
+            column_name, primary_key_column.type
+        )
+        op.add_column(referrer_table.name, column, schema=schema)
+        if unique_link:
+            op.create_unique_constraint(None, referrer_table.name, [column_name], schema=schema)
+        op.create_foreign_key(
+            None,
+            referrer_table.name,
+            referent_table.name,
+            [column.name],
+            [primary_key_column.name],
+            source_schema=schema,
+            referent_schema=schema
+        )
+
+
+def create_many_to_many_link(engine, schema, map_table_name, referents):
+    with engine.begin() as conn:
+        referent_tables_oid = [referent['referent_table'] for referent in referents]
+        referent_tables = reflect_tables_from_oids(referent_tables_oid, engine, conn)
+        metadata = MetaData(bind=engine, schema=schema, naming_convention=naming_convention)
+        opts = {
+            'target_metadata': metadata
+        }
+        ctx = MigrationContext.configure(conn, opts=opts)
+        op = Operations(ctx)
+        op.create_table(map_table_name, schema=schema)
```

Comment:
I think you should use the standard Mathesar function for creating a table. If nothing else, we'll want the default columns so this fits into the rest of the framework, correct?

Reply:
`create_mathesar_table` doesn't respect the transaction. The below test case fails due to `create_table` not being part of the transaction and ends up creating the tables although the foreign key constraint step fails. I am not sure if I am doing something wrong 

```
def test_table_creation(engine_with_schema):
    engine, schema = engine_with_schema
    supported_types = get_supported_alter_column_types(
        engine, friendly_names=False,
    )
    sa_type = supported_types.get("INTEGER")
    metadata = MetaData(bind=engine, schema=schema, naming_convention=naming_convention)
    with engine.begin() as conn:
        metadata.reflect()
        opts = {
            'target_metadata': metadata
        }
        ctx = MigrationContext.configure(conn, opts=opts)
        referent_column = MathesarColumn(
            "reference",
            sa_type()
        )
        target_table = create_mathesar_table(
            "target_table",
            schema,
            [referent_column],
            engine,
            metadata,
        )
        fk_column = MathesarColumn(
            "fk_col",
            sa_type()
        )
        base_table = create_mathesar_table(
            "base_table",
            schema,
            [fk_column],
            engine,
            metadata
        )
        op = Operations(ctx)
        # op.create_unique_constraint(None, target_table.name, [referent_column.name], schema=schema)
        try:
            op.create_foreign_key(
                None, base_table.name,
                target_table.name,
                [fk_column.name], [referent_column.name], source_schema=schema, referent_schema=schema
            )
        except:
            pass
    schema_obj = reflect_schema(engine, schema)
    tables = get_table_oids_from_schema(schema_obj.oid, engine)
    # Table should not be created as foreign key creation step would have failed due to referent column missing the unique constraint 
    assert len(tables) == 0
```

Reply:
The transaction fails if `create_mathesar_table` uses a connection to bind instead of engine when creating the table 

https://github.com/centerofci/mathesar/blob/b2467dcf63de786c0e8063821ff317ef608ec3a6/db/tables/operations/create.py#L28
~table.create(engine)~
table.create(connection)

Reply:
Argh, what a pain.

I think it would still be better to use the `create_mathesar_table` function to create the table, since it's important that we have the pieces needed (`id` column at a minimum). We should then file a bug to deal with the fact that transactions aren't working properly.

I really want to avoid ad-hoc table creation calls scattered through the codebase since we already have a default column involved and we're going to add more default metadata, etc. over time.